Add AI Prompt

bact opened the pull request:
Based on 1 Oct 2025 AI WG meeting, this PR includes changes needed to supporting AI prompt.

It incorporates some information from the model diagram at https://lists.spdx.org/g/spdx-ai/message/338 and @ElyasRashno PRs:

- #1090 -> `/Core/modality`
- #1078 -> `/Core/Modality`
- #1085 -> `/AI/isContextAugmented`
- #1082 -> `/AI/promptPattern`

Prompt chaining can be described by using the new "follows" relationship type, see:

- #1116

**Note that more properties can be added later, this is just a skeleton.**

<img width="608" height="514" alt="ai-prompt" src="https://github.com/user-attachments/assets/0f2fa262-974c-4d11-b91e-b567df8f13e4" />

> When `isContextAugmented` is set to `true`, a Relationship of type `usesTool`
  can be optionally used to describe the context augmentation mechanism or tool
  employed (e.g., referencing a retrieval-augmented generation tool).

```mermaid
graph LR;
  B(Prompt) -->|usesTool| A(RAG)
```

## Design notes

There are few redesigns (see notes below) and revisions in class/property definitions, please look at the changes in each individual file.

1) Subclassing **`/AI/Prompt`** from `/Software/SoftwareArtifact`:

    - It represents a reasonable unit of content that can be attributed to a specific source.
    - It is lighter than a `/Dataset/DatasetPackage` and requires less modification to the external properties restrictions.
    - Reuse `/Software/contentIdentifier` for prompt content URI. The type of content identifier URI can be either "Gitoid" or "SWHID" (if needed, a new content identifier type can be added to `/Software/ContentIdentifierType`).

2) The **`/Core/contentModality`** (property) and **`/Core/Modality`** (vocab) are trying to be generic enough for use cases in foundation model and in human-computer interaction.
    - Note that modality is more about mode of communication/interaction, and less about data format, but it's not 100% dichotomy here (see also [Hugging Face modalities](https://huggingface.co/datasets), which is also mixed).
    - Alternatively, we can adopt a much simpler vocab from Schema.org. Schema.org [`accessMode`](https://schema.org/accessMode) has 4 approved values: "auditory", "tactile", "textual", "visual". See: [Schema.org Accessibility Properties for Discoverability Vocabulary](https://www.w3.org/community/reports/a11y-discov-vocab/CG-FINAL-vocabulary-20241209/#accessMode-vocabulary).
    - In any case, we can make a decision to also include [`/Core/contentType`](https://github.com/spdx/spdx-3-model/blob/develop/model/Core/Properties/contentType.md) which indicate media type of a content. Since it is based on MIME type, it is limited, but should cover general use cases (like differentiate still image from moving image, or tell if the "textual" input is actually a CSV data).

3) Use **`isContextAugmented`** name, instead of `hasRAG`, as we may like to focus on the inclusion of automatically augmented context/supporting content to the prompt content, rather than the technique (RAG, etc).

4) Some existing definitions are slightly revised to make it accommodate more use cases or make it more precise. For example:

    - In `/Dataset/dataPreprocessing`, use "input data" instead of "raw data", since the data to be preprocessed could be data from other previous stages in the pipeline and not necessarily raw data.
    - In `/Dataset/dataCollectionProcess`, use "data" instead of "dataset". This will be more precise to the property name and make it more more naturally applied to a prompt.
        - We may also need to explain what "data collection" means here for prompt. Does it mean how a prompt is constructed? If not relevant, we can remove this field.

bact committed:
Add AI Prompt

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>
Co-Authored-By: ElyasRashno <42498039+elyasrashno@users.noreply.github.com>
Co-Authored-By: stevenc-stb <192034142+stevenc-stb@users.noreply.github.com>

bact committed:
modality -> /Core/modality

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
ModalityType -> Modality

Moved to vocab

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Remove hasInputOutputExample for now

The property is intended to indicate if the prompt contains few-shot examples or not. But remove it for now.

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
hasSensitivePersonalInformation -> /Dataset/hasSensitivePersonalInformation

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Fix range

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Make data* properties more generic

So the dataCollectionProcess and dataPreprocessing properties can be used for any kind of data, including prompts.

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Reuse contentIdentifier for prompt content

Reuse contentIdentifier for prompt content URI, which can be either Gitoid or SWHID.

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Fix external properties

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
raw data -> input data

The data to be preprocessed could be data from other previous stages in the pipeline and not necessarily raw data.

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Make plural in contentIdentifier desc

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
hasRAGContext -> isContextAugmented

Use a more tech-neutral term. Ref: https://learn.microsoft.com/en-us/azure/ai-foundry/openai/concepts/prompt-engineering#supporting-content

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Add usesTool

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Add Prompt diagram

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>
Co-Authored-By: stevenc-stb <192034142+stevenc-stb@users.noreply.github.com>
Co-Authored-By: ElyasRashno <42498039+elyasrashno@users.noreply.github.com>

bact committed:
Update diagram

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Fix typo in dataPreprocessing attribute

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact commented:
From 8 Oct 2025 AI WG meeting, need a definition of

- [x] prompt
- [ ] context

bact commented:
Follow-up of the discussion from 8 Oct 2025 AI WG meeting on defining ordinality between two prompts (Prompt1 preceeds Prompt2), a possibility of having a "follows" relationship type is proposed to model a sequence of prompts.

See the "follows" PR: 
- https://github.com/spdx/spdx-3-model/pull/1116

goneall commented:
@bact @rgopikrishnan91 - It looks like this is still under discussion.  Should this be a draft PR?  Perhaps the AI team could review the open PR's and update with comments and/or close and/or merge.

bact commented:
Thanks. We will put all AI-related PRs that still under discussion to draft first.

ElyasRashno commented on model/AI/Classes/Prompt.md:
```diff
@@ -0,0 +1,54 @@
+
+## Summary
+
+Specifies the content of a prompt and its associated information.
```
Specifies the prompt content and its associated information.

ElyasRashno commented on model/AI/Classes/Prompt.md:
```diff
@@ -0,0 +1,54 @@
+- Instantiability: Concrete
+
+## Properties
+
```
- /Core/autonomyType
   - type: /Core/PresenceType
   - minCount: 0

bact replied:
An agent can have autonomy/automation level, as it can take action on its own.
Not sure if a prompt can have the autonomy/automation level, as it is more of a piece of communication between agents (human and machine) ?

ElyasRashno commented on model/AI/Classes/Prompt.md:
```diff
@@ -0,0 +1,54 @@
+- isContextAugmented
+  - type: xsd:boolean
+  - minCount: 0
+  - maxCount: 1
```
- promptPattern
   - type: xsd:string
   - minCount: 0

bact replied:
Added based on summary and description from https://github.com/spdx/spdx-3-model/pull/1082

ElyasRashno commented:
@bact, I’ve added some comments with change suggestions in the prompt class. I also need access to this PR to create the corresponding vocabulary.

bact committed:
Make summary more concise

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>
Co-Authored-By: ElyasRashno <42498039+elyasrashno@users.noreply.github.com>

bact committed:
Add Elyas suggestions

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>
Co-Authored-By: ElyasRashno <42498039+elyasrashno@users.noreply.github.com>

bact committed:
Expand beyond reasoning

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>
Co-Authored-By: ElyasRashno <42498039+elyasrashno@users.noreply.github.com>

bact commented:
> I’ve added some comments with change suggestions in the prompt class. I also need access to this PR to create the corresponding vocabulary.

@ElyasRashno Thank you - incorporated. <s>Will update the diagram.</s> (updated)

GitHub has no facility to share access to an individual PR. Please instead create a PR that targets the branch of this PR (https://github.com/bact/spdx-3-model/tree/add-ai-prompt). Like the example that I have emailed you. Or you can also use the "Add a suggestion" button when commenting on code.

<img width="774" height="257" alt="github-add-a-suggestion" src="https://github.com/user-attachments/assets/9d94e887-ae15-4b5c-b7d0-2013783935c5" />

bact committed:
Update diagram

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>
Co-Authored-By: ElyasRashno <42498039+elyasrashno@users.noreply.github.com>

bact committed:
Add prompt definition

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Replace prompt modality with contentModality

Be more specific.
Also revise Modality vocab to be a simple one.
Complex data types should more belong to /Dataset/DatasetType vocab.

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Update diagram

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Remove change in ContentIdentifier

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Add context definition

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Update with ISO verbal forms

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Update inLanguage to match recent changes

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact commented:
For current definitions of "prompt" and "context", see in the files:
https://github.com/spdx/spdx-3-model/pull/1110/files

> A prompt is the input text, instruction, or data given to an AI model to
initiate and guide its generated output.

> Context is the information provided to the model that shapes or influences
  its generated output.

bact commented:
From today's Prompt/Context meeting (2025-10-17):

**ONLY FOR MEETING PARTICIPANTS UNDERSTANDING DURING DISCUSSION**

<img width="1441" height="822" alt="56443822-4230-47A3-8443-45D4CEDE57FA" src="https://github.com/user-attachments/assets/48af8aa4-006e-41c8-a02b-5270a39f905c" />

<img width="411" height="392" alt="99D0854D-11BB-4A41-9B10-963D1977D742" src="https://github.com/user-attachments/assets/b3510e04-b186-4789-9f25-1b79831601a0" />

<img width="839" height="401" alt="713F7297-C704-4955-BA91-BDE98556A545" src="https://github.com/user-attachments/assets/8e23a105-ab95-4869-ad4f-ff65f0656c37" />

bact committed:
Merge branch 'spdx:develop' into add-ai-prompt

bact committed:
Merge branch 'spdx:develop' into add-ai-prompt

kestewart commented:
Please let us know if this won't make 3.1 and should be moved to 3.2

bact commented:
With the current state of direction within the AI WG, the fundamental design has changed.

We can close this and a new PR with the updated design can be opened later.
